pkg/engine/internal/workflow/workflow.go

```diff
@@ -133,19 +144,46 @@ func (wf *Workflow) Run(ctx context.Context) (pipeline executor.Pipeline, err er
 		}
 	}
 
-	// TODO(rfratto): For logs queries, we want a system to limit how many scan
-	// tasks get sent to the runner at once.
-	//
-	// This will limit unnecessary resource consumption of workflows when
-	// there's a lot of compute capacity.
-	if err := wf.runner.Start(ctx, wrappedHandler, tasks...); err != nil {
-		pipeline.Close()
+	g := wf.dispatchTasks(ctx, wrappedHandler, tasks)
+	if err := g.Wait(); err != nil {
```

Comment:
I believe calling `Wait` here may be prone to deadlocks.

For example, let's say we have 100 scan tasks, but can only schedule 10 at a time. Eventually, those scan tasks will hit backpressure sending results to the final stream, because nothing is reading from `wrapped` (as we're blocked on waiting for Wait before returning `wrapped` back to the caller).

Reply:
Looks like the order in which the tasks are enqueued matters. If the scan tasks are scheduled first and then the remaining, then it blocks.

Reply:
This happens due to "worker starvation" where there are not enough workers any more to run tasks that are required to resolved the dependency tree of the execution graph.